destination/destination.go

```diff
@@ -0,0 +1,198 @@
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/conduitio-labs/conduit-connector-snowflake/destination/format"
+	"github.com/conduitio-labs/conduit-connector-snowflake/repository"
+	sdk "github.com/conduitio/conduit-connector-sdk"
+	"github.com/go-errors/errors"
+)
+
+const (
+	defaultBatchDelay = time.Second * 5
+	defaultBatchSize = 1000
+)
+
+type Destination struct {
+	sdk.UnimplementedDestination
+	repository *repository.Snowflake
+	config     Config
+}
+
+// NewDestination creates the Destination and wraps it in the default middleware.
+func NewDestination() sdk.Destination {
+	// This is needed to override the default batch size and delay defaults for this destination connector.
+	middlewares := sdk.DefaultDestinationMiddleware()
+	for i, m := range middlewares {
+		switch dest := m.(type) {
+		case sdk.DestinationWithBatch:
+			dest.DefaultBatchDelay = defaultBatchDelay
+			dest.DefaultBatchSize = defaultBatchSize
+			middlewares[i] = dest
+		}
+	}
+	return sdk.DestinationWithMiddleware(&Destination{}, middlewares...)
+}
+
+func (d *Destination) Parameters() map[string]sdk.Parameter {
+	return Config{}.Parameters()
+}
+
+func (d *Destination) Configure(ctx context.Context, cfg map[string]string) error {
+	sdk.Logger(ctx).Debug().Msg("Configuring Destination Connector.")
+
+	// TODO: add configuration for ordering column (aka primary key)
+	// right now, we will automatically detect the key from the key within the record,
+	// but it would be great to have flexibility in case the user wants to key on a different
+
+	err := sdk.Util.ParseConfig(cfg, &d.config)
+	if err != nil {
+		return fmt.Errorf("failed to parse destination config : %w", err)
+	}
+
+	return nil
+}
+
+func (d *Destination) Open(ctx context.Context) error {
+	// Open is called after Configure to signal the plugin it can prepare to
+	// start writing records. If needed, the plugin should open connections in
+	// this function.
+	var err error
+	d.repository, err = repository.Create(ctx, d.config.Connection)
+	if err != nil {
+		return fmt.Errorf("failed to create snowflake repository: %w", err)
+	}
+
+	return nil
+}
+
+func (d *Destination) Write(ctx context.Context, records []sdk.Record) (int, error) {
+	// TLDR - we don't need to implement custom batching logic, it's already handled
+	// for us in the SDK, as long as we use sdk.batch.size / sdk.batch.delay.
+
+	// sdk.batch.size and sdk.batch.delay already handles batching and should
+	// control the size of records & timing of when Write() method is invoked.
+	// FYI - these are only implemented in the SDK for destinations
+
+	// General approach
+
+	// TODO: move this into another package.
+
+	// TODO: read naming prefix from config
+
+	csv, schema, cols, orderingCols, err := format.MakeCSVRecords(records, "meroxa")
+	if err != nil {
+		return 0, errors.Errorf("failed to convert records to CSV: %w", err)
+	}
+
+	file, err := os.CreateTemp(os.TempDir(), "snowflake_*.csv")
```

Comment:
@samirketema the idea of using a steam is to pass an `io.Reader`, which means we could just use a memory buffer, so no temp files are created.

Reply:
gotcha - will update this soon. thanks.